libtiledbsoma/src/soma/soma_array.h

```diff
@@ -775,6 +847,108 @@ class SOMAArray : public SOMAObject {
+        } else {
+            return _get_core_domain();
+        }
+    }
+
+    /**
+     * Returns the SOMA maxdomain in its entirety, as an Arrow table for return
+     * to Python/R.
+     *
+     * o For arrays with core current-domain support:
+     *   - soma domain is core current domain
+     *   - soma maxdomain is core domain
+     * o For arrays without core current-domain support:
+     *   - soma domain is core domain
+     *   - soma maxdomain is core domain
+     *   - core current domain is not accessed at the soma level
+     *
+     * @tparam T Domain datatype
+     * @return Pair of [lower, upper] inclusive bounds.
+     */
+    ArrowTable get_soma_maxdomain() {
+        return _get_core_domain();
+    }
+
+    /**
+     * Returns the core non-empty domain in its entirety, as an Arrow
+     * table for return to Python/R.
+     */
+    ArrowTable get_non_empty_domain() {
+        return _get_core_domainish(Domainish::kind_non_empty_domain);
+    }
+
+    /**
+     * Code-dedupe helper for core domain, core current domain, and core
+     * non-empty domain.
+     */
+    ArrowTable _get_core_domainish(enum Domainish which_kind);
+
+    /**
+     * This enables some code deduplication between core domain, core current
+     * domain, and core non-empty domain.
+     */
+    template <typename T>
+    std::pair<T, T> _core_domainish_slot(
+        const std::string& name, enum Domainish which_kind) const {
+        if (std::is_same_v<T, std::string>) {
+            throw std::runtime_error(
+                "SOMAArray::_core_domainish_slot: template-specialization "
+                "failure.");
+        }
+        switch (which_kind) {
+            case Domainish::kind_core_domain:
+                return _core_domain_slot<T>(name);
+                break;
+            case Domainish::kind_core_current_domain:
+                return _core_current_domain_slot<T>(name);
+                break;
+            case Domainish::kind_non_empty_domain:
+                return non_empty_domain_slot<T>(name);
+                break;
+            default:
+                throw std::runtime_error(
+                    "internal coding error in SOMAArray::_core_domainish_slot");
+        }
+    }
+
+    std::pair<std::string, std::string> _core_domainish_slot_string(
+        const std::string& name, enum Domainish which_kind) const {
+        switch (which_kind) {
+            case Domainish::kind_core_domain:
+                return _core_domain_slot_string(name);
+                break;
+            case Domainish::kind_core_current_domain:
+                return _core_current_domain_slot<std::string>(name);
+                break;
+            case Domainish::kind_non_empty_domain:
+                return non_empty_domain_slot_var(name);
+                break;
+            default:
+                throw std::runtime_error(
+                    "internal coding error in SOMAArray::_core_domainish_slot");
```

Comment:
Do we also want to elaborate that the internal coding error is an unknown `Domainish` type?

```suggestion
                    "internal coding error in SOMAArray::_core_domainish_slot_string");
```

Reply:
good call! :)